node/src/p2p/swarm_manager.rs

```diff
@@ -0,0 +1,703 @@
+            })
+            .unwrap_or(false);
+
+        if kad_entry_exists {
+            // When `connect` is called without a specified address, then
+            // swarm asks Kademlia about it.
+            self.connect(peer_id, None);
+            return;
+        }
+
+        // When Kademlia doesn't know about peer_id, then we need to
+        // initiate a `get_closest_peers` query in order to find it and
+        // their addresses.
+        //
+        // We also avoid calling `get_closest_peers` for the same peer
+        // multiple times.
+        let peer_id_bytes = peer_id.to_bytes();
+        let kad_query_exists = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .iter_queries()
+            .any(|query| match query.info() {
+                QueryInfo::GetClosestPeers { key, .. } => *key == peer_id_bytes,
+                _ => false,
+            });
+
+        if !kad_query_exists {
+            // When kademlia finds the addresses via `get_closest_peers`, it will
+            // automatically connect to them.
+            self.swarm
+                .behaviour_mut()
+                .kademlia
+                .get_closest_peers(*peer_id);
+        }
+    }
+
+    fn connect_to_bootnodes(&mut self) {
+        // Collect all the bootnodes that are not currently connected.
+        let bootnodes = self
+            .bootnodes
+            .iter()
+            .filter_map(|(peer_id, addrs)| {
+                if self.peer_tracker.is_connected(peer_id) {
+                    None
+                } else {
+                    Some((peer_id.to_owned(), addrs.to_owned()))
+                }
+            })
+            .collect::<Vec<_>>();
+
+        if bootnodes.is_empty() {
+            return;
+        }
+
+        // We produce this event only if we are going to connect to at least
+        // one bootnode.
+        self.event_pub.send(NodeEvent::ConnectingToBootnodes);
+
+        for (peer_id, addrs) in bootnodes {
+            for addr in &addrs {
+                self.swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .add_address(&peer_id, addr.to_owned());
+            }
+
+            self.connect(&peer_id, addrs);
+        }
+    }
+
+    fn bootstrap(&mut self) {
+        self.connect_to_bootnodes();
+
+        // Check if there is bootstrap in progress
+        let bootstrap_query_exists = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .iter_queries()
+            .any(|query| matches!(query.info(), QueryInfo::Bootstrap { .. }));
+
+        if !bootstrap_query_exists {
+            if let Err(e) = self.swarm.behaviour_mut().kademlia.bootstrap() {
+                warn!("Can't run kademlia bootstrap: {e}");
+            }
+        }
+    }
+
+    fn start_get_providers_kad_query(&mut self, topic: &RecordKey) {
+        // Avoid having multiple `GetProviders` queries for the same `topic`.
```

Comment:
```suggestion
        // Avoid running multiple `GetProviders` queries for the same `topic`.
```
nit, I felt it more to the point here, since it's specificially about concurrently running queries